Use our own implementation of miekg/dns.TsigProvider interface

irbekrm opened the pull request:
See https://github.com/cert-manager/cert-manager/pull/4942 for context

This PR adds our own implementation of [github.com/miekg/dns.TsigProvider interface](https://github.com/miekg/dns/blob/33e64002b6f9bc4d1c694ce6677d6185ba37544f/tsig.go#L28)

This is done to allow us to both upgrade the upstream library and keep supporting HMACMD5 as RFC2136 TSIG algorithm which was deprecated in the upstream library

```release-note
NONE
```

/kind cleanup

Signed-off-by: irbekrm <irbekrm@gmail.com>

wallrj commented on pkg/issuer/acme/dns/rfc2136/tsig.go:
```diff
@@ -0,0 +1,97 @@
+	return nil
+}
+
+func fromBase64(s []byte) ([]byte, error) {
+	buflen := base64.StdEncoding.DecodedLen(len(s))
+	buf := make([]byte, buflen)
+	n, err := base64.StdEncoding.Decode(buf, s)
+	buf = buf[:n]
+	return buf, err
+}
```
https://github.com/miekg/dns/blob/7318b01e11efedb3dda0453a0ccbd1fd42b062ed/msg_helpers.go#L165-L171

is subtly different (naked return) but looks functionally identical.

irbekrm replied:
yeah it is, I removed the naked return as it differs from our coding style

wallrj commented on pkg/issuer/acme/dns/rfc2136/tsig.go:
```diff
@@ -0,0 +1,97 @@
+	return h.Sum(nil), nil
+}
+
+// Verify is copied from github.com/miekg/dns.tsigHMACProvider.Verify
```
Might be nice to link to the exact Git hash where this was copied from, so that if it gets moved upstream, we can more easily locate the original and see how it's changed since we last copied.

wallrj commented on pkg/issuer/acme/dns/rfc2136/tsig.go:
```diff
@@ -0,0 +1,97 @@
+		return dns.ErrSig
+	}
+	return nil
+}
```
I tried to think if there was some way we could avoid copying this code, but nothing comes to mind.
I see that we're not testing this, but I see that upstream do have some tests for this.

irbekrm replied:
Yes, if the upstream implementation of `TsigProvider` was exported, we could probably use the `Verify` method from the upstream type, but I couldnt' think of the way how to do it now.

wallrj commented on pkg/issuer/acme/dns/rfc2136/tsig.go:
```diff
@@ -0,0 +1,97 @@
+var _ dns.TsigProvider = tsigHMACProvider("")
+
+// Generate is a largely copied from
+// github.com/miekg/dns.tsigHMACProvider.Generate except for supporting HMACMD5
```
Maybe link to https://github.com/miekg/dns/blob/d70eb7b9e19e2d91e54e655209bdb262c735e6bf/tsig.go#L37-L60 so that it's easy to know exactly what we copied.

irbekrm commented on pkg/issuer/acme/dns/rfc2136/tsig_test.go:
```diff
@@ -0,0 +1,159 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_tsigHMACProvider_Generate(t *testing.T) {
```
Note: this unit test is largely just testing that the upstream libaries work as expected, but would catch a very basic error i.e in the switch and serves as a very basic documentation for what the Generate method does.

wallrj replied:
Yeah, I wasn't sure it was necessary. But I did check the coverage which seems fine.

![image](https://user-images.githubusercontent.com/978965/159032491-886b1a5a-ffbc-47c5-838c-bcbdddddb2a0.png)

wallrj commented on pkg/issuer/acme/dns/rfc2136/tsig_test.go:
```diff
@@ -0,0 +1,159 @@
+			if (err != nil) != scenario.wantErr {
+				t.Errorf("tsigHMACProvider.Generate() error = %v, wantErr %v", err, scenario.wantErr)
+			}
+			if !reflect.DeepEqual(got, scenario.want) {
+				t.Errorf("tsigHMACProvider.Generate() = %v, want %v", got, scenario.want)
+			}
```
Why use `reflect.DeepEqual` for comparing byte slices? Wouldn't `bytes.Equal` be more appropriate? 
And `assert.Equal` might be even better and give readable output if the bytes differ.

irbekrm replied:
Thanks, that's definitely a better way to do it 👍🏼

wallrj reviewed (approved):
Thanks @irbekrm 

It's a shame that upstream can't be persuaded to just re-instate the MD5 hmac feature.
I see that others are advocating for it. Perhaps we should too.

This all looks good to me:
 * I suggested a few links to the exact upstream source code.
 * Perhaps strictly we should include an extra copyright comment acknowledging the upstream copyright, but it's not clear which https://github.com/miekg/dns/blob/master/LICENSE or https://github.com/miekg/dns/blob/master/COPYRIGHT or all of those.


/lgtm
/hold in case you want to address those comments.

irbekrm commented on pkg/issuer/acme/dns/rfc2136/tsig.go:
```diff
@@ -0,0 +1,97 @@
+	default:
+		return nil, dns.ErrKeyAlg
+	}
+	_, err = h.Write(msg)
+	if err != nil {
+		return nil, err
+	}
```
I assumed they have forgotten to check for it

irbekrm commented on pkg/issuer/acme/dns/rfc2136/tsig.go:
```diff
@@ -0,0 +1,97 @@
+// github.com/miekg/dns.tsigHMACProvider.Generate except for supporting HMACMD5
+func (key tsigHMACProvider) Generate(msg []byte, t *dns.TSIG) ([]byte, error) {
+	rawsecret, err := fromBase64([]byte(key))
+	if err != nil {
+		return nil, fmt.Errorf("error decoding the provided TSIG secret: %w", err)
+	}
```
> Why not stay identical to upstream except for the added case statement below

It's very similar to upstream at the moment, I made a few XS changes besides adding the case statement. I don't really want to add a comment saying it's identical as there is no guarantee that someone doesnt' change the code at a later point even more, but doesn't update the comment- but I will link to the exact comment as you suggested, so a future reader can compare for themselves

irbekrm commented:
Thanks for the thorough code review @wallrj , I think I've addressed all the points, PTAL

> It's a shame that upstream can't be persuaded to just re-instate the MD5 hmac feature.
I see that others are advocating for it. Perhaps we should too.

I will leave a note to the issue (and will leave a link to this PR in case it makes it easier for anyone else trying to figure out how to implement the interface)

> Perhaps strictly we should include an extra copyright comment acknowledging the upstream copyright, but it's not clear which https://github.com/miekg/dns/blob/master/LICENSE or https://github.com/miekg/dns/blob/master/COPYRIGHT or all of those.

Have we done this before? I know there are other bits in the codebase where the code is copied over from kubernetes or other upstream projects

wallrj commented:
/retest
/lgtm
/unhold

wallrj commented:
> Have we done this before? I know there are other bits in the codebase where the code is copied over from kubernetes or other upstream projects

Here for example

 https://github.com/cert-manager/cert-manager/blob/c0da0894ba29efb3d4c948b974c33100a1637ace/pkg/webhook/admission/initializer/initializer.go#L17-L20

See https://github.com/cert-manager/cert-manager/pull/4664#discussion_r780241280

irbekrm commented:
@wallrj could I get another LGTM please?

Looks like I had to rebase the PR to get rid of the (unrelated) flakes.

wallrj commented:
/lgtm

wallrj commented:
/lgtm
